File: client/surbased/lib/src/survey/domain/survey_model.dart
import 'dart:convert';

import 'package:surbased/src/survey/domain/question_model.dart';
import 'package:surbased/src/survey/domain/tag_model.dart';
import 'package:surbased/src/user/domain/user_model.dart';

Survey surveyFromJson(String str) => Survey.fromJson(json.decode(str));

String surveyToJson(Survey data) => json.encode(data.toJson());

class Survey {
  String name;
  String? assignmentStatus;
  int? invitationsRejected;
  String categoryId;
  String ownerId;
  final String? id;
  String? description;
  String scope;
  DateTime? startDate;
  DateTime? endDate;
  final List<Question> questions;
  List<User>? assignedUsers;
  String? organizationId;
  List<Tag>? tags;
  int? responseCount;

  Survey({
    required this.name,
    this.assignmentStatus,
    this.invitationsRejected,
    required this.categoryId,
    required this.ownerId,
    this.id,
    this.description,
    required this.scope,
    required this.startDate,
    required this.endDate,
    required this.questions,
    this.assignedUsers,
    this.organizationId,
    this.tags,
    this.responseCount,
  });

  factory Survey.fromJson(Map<String, dynamic> json) => Survey(
        name: json["name"] ?? '',
        assignmentStatus: json["assignment_status"],
        invitationsRejected: json["invitations_rejected"],
        categoryId: json["category_id"] ?? '',
        ownerId: json["owner_id"] ?? '',
        id: json["id"],
        description: json["description"] ?? '',
        scope: json["scope"] ?? 'private',
        organizationId: json["organization_id"],
        startDate: DateTime.parse(json["start_date"]),
        endDate: DateTime.parse(json["end_date"]),
        questions: json["questions"] == null 
            ? [] 
            : List<Question>.from(json["questions"].map((x) => Question.fromJson(x))),
        tags: json["tags"] == null 
            ? [] 
            : List<Tag>.from(json["tags"].map((x) => Tag.fromJson(x))),
        responseCount: json["response_count"],
      );

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> json = {};
    json["name"] = name;
    if (assignmentStatus != null) json["assignment_status"] = assignmentStatus;
    if (invitationsRejected != null) json["invitations_rejected"] = invitationsRejected;
    json["category_id"] = categoryId;
    json["owner_id"] = ownerId;
    if (id != null) json["id"] = id;
    if (description != null) json["description"] = description;
    json["scope"] = scope;
    if (organizationId != null) json["organization_id"] = organizationId;
    if (startDate != null) json["start_date"] = '${startDate?.year}-${startDate?.month.toString().padLeft(2, '0')}-${startDate?.day.toString().padLeft(2, '0')}';
    if (endDate != null) json["end_date"] = '${endDate?.year}-${endDate?.month.toString().padLeft(2, '0')}-${endDate?.day.toString().padLeft(2, '0')}';
    json["questions"] = List<dynamic>.from(questions.map((x) => x.toJson()));
    if (assignedUsers != null) json["assigned_users"] = List<dynamic>.from(assignedUsers!.map((x) => x.toJson()));
    if (tags != null) json["tags"] = List<dynamic>.from(tags!.map((x) => x.toJson()));
    if (responseCount != null) json["response_count"] = responseCount;
    return json;
  }
}

File: api/src/models/SurveyModel.py
from typing import List, Optional, TYPE_CHECKING
import uuid
from src.models.TagModel import Tag
from src.database import Base
from sqlalchemy import CheckConstraint, Column, ForeignKey, Integer, String, Date, UniqueConstraint, func, UUID
from sqlalchemy.ext.hybrid import hybrid_property
from datetime import date
from sqlalchemy.orm import  Mapped, mapped_column, relationship
from src.models.SurveyUserModel import survey_user
from src.models.SurveyTagModel import survey_tag


if TYPE_CHECKING:
    from src.models.OrganizationModel import Organization
    from src.models.UserModel import User
    from src.models.CategoryModel import Category
    from src.models.QuestionModel import Question

class Survey(Base):
    __tablename__ = "surveys"
 

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, index=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str] = mapped_column(String(250), nullable=False)
    scope: Mapped[str] = mapped_column(String(100), nullable=False)
    start_date: Mapped[date] = mapped_column(Date, nullable=False) 
    end_date: Mapped[Optional[date]] = mapped_column(Date, nullable=False)
    owner_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("users.id"), nullable=False)
    organization_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("organizations.id"), nullable=True)
    category_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("categories.id"), nullable=False)

 

    assigned_users: Mapped[Optional[List["User"]]] = relationship(secondary=survey_user, back_populates="surveys_assigned",lazy="selectin")
    owner: Mapped["User"] = relationship(back_populates="surveys_owned", lazy="selectin")
    category: Mapped["Category"] = relationship(back_populates="surveys", lazy="selectin")
    questions: Mapped[Optional[List["Question"]]] = relationship(back_populates="survey", cascade="all, delete", lazy="selectin")
    organization: Mapped[Optional["Organization"]] = relationship(back_populates="surveys", lazy="selectin")
    tags: Mapped[Optional[List["Tag"]]] = relationship(secondary=survey_tag, back_populates="surveys", lazy="selectin")


    __table_args__ = (
        CheckConstraint("scope IN ('private', 'organization', 'public')", name="scope_check"),
        CheckConstraint("(scope = 'organization' AND 'organization_id' IS NOT NULL) OR (scope <> 'organization')", name="check_org_scope"),
        CheckConstraint("end_date >= start_date", name="end_date_check"),
    )

